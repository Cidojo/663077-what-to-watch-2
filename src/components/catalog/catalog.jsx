import * as React from 'react';
import * as PropTypes from 'prop-types';
import {MovieCardThumbnail} from './../movie-card-thumbnail/movie-card-thumbnail.jsx';
import {movieCardPropTypes} from './../../global-custom-types';

const ON_THUMBNAIL_MOUSE_ENTER_DELAY = 1000;

const Catalog = (props) => {
  let timerId = null;

  const {
    active,
    movieCards,
    onActiveChange,
    maxCatalogCards
  } = props;

  const _handleCardMouseEnter = (id) => {
    timerId = setTimeout(() => onActiveChange(id), ON_THUMBNAIL_MOUSE_ENTER_DELAY);
  };

  const _handleCardMouseLeave = () => {
    clearTimeout(timerId);

    onActiveChange(null);
  };

  return (
    <div className="catalog__movies-list">
      {
        movieCards
          .slice(0, maxCatalogCards).map((card) => {
            return (
              <MovieCardThumbnail
                key={`${card.id}`}
                card={card}
                isPlaying={active === card.id}
                isStopped={!(active === card.id)}
                onThumbnailMouseEnter={_handleCardMouseEnter}
                onThumbnailMouseLeave={_handleCardMouseLeave}
                onThumbnailClick={onActiveChange}
              />
            );
          })
      }
    </div>
  );
};

Catalog.propTypes = {
  active: PropTypes.number,
  movieCards: PropTypes.arrayOf(movieCardPropTypes),
  maxCatalogCards: PropTypes.number.isRequired,
  onActiveChange: PropTypes.func
};

Catalog.defaultProps = {
  active: 0,
  movieCards: [],
  maxCatalogCards: 0,
  onActiveChange: () => {}
};

export {Catalog};
