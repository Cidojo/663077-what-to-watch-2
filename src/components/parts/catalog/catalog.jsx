import * as React from 'react';
import * as PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {SmallMovieCard} from '../small-movie-card/small-movie-card.jsx';
import {ShowMoreButton} from './../show-more-button/show-more-button.jsx';

import {DisplayCountActionCreator} from './../../../reducers/display-count-reducer/display-count-reducer.js';
import {movieCardPropTypes} from './../../../global-custom-types.js';
import Selectors from './../../../selectors/selectors.js';
import {SHOW_MORE_STEP} from './../../../constants/constants.js';

const Catalog = (props) => {
  const {
    movieCards,
    active,
    withShowMoreButton,
    displayCount,
    onShowMore
  } = props;

  const handleShowMore = () => {
    onShowMore(displayCount);
  };

  return (
    <>
      <div className="catalog__movies-list">
        {movieCards.slice(0, displayCount).map((card) => {
          return <SmallMovieCard key={card.id} card={card} isActive={active === card} />;
        })}
      </div>
      {withShowMoreButton && movieCards.length >= displayCount &&
        <ShowMoreButton
          onButtonClick={handleShowMore}
        />
      }
    </>
  );
};

Catalog.propTypes = {
  active: PropTypes.number,
  movieCards: PropTypes.arrayOf(movieCardPropTypes),
  displayCount: PropTypes.number,
  withShowMoreButton: PropTypes.bool,
  onShowMore: PropTypes.func.isRequired
};

Catalog.defaultProps = {
  active: null,
  movieCards: [],
  withShowMoreButton: false,
  displayCount: 0,
  onShowMore: () => {}
};

const mapDispatchToProps = (dispatch) => ({
  onShowMore: (shownCount) => {
    dispatch(DisplayCountActionCreator.incrementCount(shownCount + SHOW_MORE_STEP));
  }
});

const mapStateToProps = (state, ownProps) => {
  return Object.assign({}, ownProps, {
    displayCount: Selectors.getDisplayCount(state)
  });
};

export {Catalog};
export default connect(mapStateToProps, mapDispatchToProps)(Catalog);
