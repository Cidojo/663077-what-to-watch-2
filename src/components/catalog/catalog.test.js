import * as React from 'react';
import {Catalog} from './catalog.jsx';
import * as renderer from 'react-test-renderer';

it(`should render component without errors`, () => {
  const tree = renderer
    .create(<Catalog
      movieCards={[{
        id: 0,
        title: ``,
        link: ``,
        imgSrc: ``,
        posterSrc: ``,
        imgDescription: ``,
        genre: ``,
        year: 0
      }]}
      cardsPerPage={0}
      handleCurrentVideoIDChange={() => {}}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
