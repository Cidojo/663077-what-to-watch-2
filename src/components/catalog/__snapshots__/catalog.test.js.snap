// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component without errors 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        className="player__video"
        poster=""
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href=""
        onClick={[Function]}
      >
        
      </a>
    </h3>
  </article>
</div>
`;
